README
MySQL/MariaDB transaction_time audit plugin
===========================================

This plugin logs long running transactions to the server
error log when transaction took more than a configurable
number of seconds.

Installation & Activation
-------------------------

See INSTALL file

Configuration variables
-----------------------

transaction_time_limit
^^^^^^^^^^^^^^^^^^^^^^

Log long running transaction to error log if it took more
than this many seconds (default=10s, disabled if set to 0s)

STATUS values
-------------

transaction_time_start
^^^^^^^^^^^^^^^^^^^^^^

Time the current transaction was started as a Unix Timestamp
(seconds since 1970), or 0 if not currently in an active 
transaction.

Note: a transaction doesn't actually start with BEGIN or
START TRANSACTION but with the first actual statement touching
data after that

transaction_time_span
^^^^^^^^^^^^^^^^^^^^^

Time in seconds since the transaction started or 0 if not in
an active transaction.



NEWS
This is a source project generated by CodeGen_MySQL_Plugin_Extension 0.9.3dev
...


myplugin_audit.h
/*
   +----------------------------------------------------------------------+
   | This program is free software; you can redistribute it and/or        |
   | modify it under the terms of the GNU General Public License          |
   | as published by the Free Software Foundation; either version 2       |                 
   | of the License, or (at your option) any later version.               | 
   |                                                                      |
   | This program is distributed in the hope that it will be useful,      |
   | but WITHOUT ANY WARRANTY; without even the implied warranty of       |
   | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        |
   | GNU General Public License for more details.                         | 
   |                                                                      |
   | You should have received a copy of the GNU General General Public    |
   | License in the file LICENSE along with this library;                 |
   | if not, write to the                                                 | 
   |                                                                      |
   |   Free Software Foundation, Inc.,                                    |
   |   51 Franklin Street, Fifth Floor,                                   |
   |   Boston, MA  02110-1301  USA                                        |
   +----------------------------------------------------------------------+
   | Authors: Hartmut Holzgraefe <hartmut@mysql.com>                      |
   +----------------------------------------------------------------------+
*/

/* $ Id: $ */ 

#ifndef MYPLUGIN_AUDIT_H
#define MYPLUGIN_AUDIT_H


#endif /* MYPLUGIN_AUDIT_H */



transaction_time_audit.c
/*
  +----------------------------------------------------------------------+
  | This program is free software; you can redistribute it and/or        |
  | modify it under the terms of the GNU General Public License          |
  | as published by the Free Software Foundation; either version 2       |
  | of the License, or (at your option) any later version.               |
  |                                                                      |
  | This program is distributed in the hope that it will be useful,      |
  | but WITHOUT ANY WARRANTY; without even the implied warranty of       |
  | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        |
  | GNU General Public License for more details.                         |
  |                                                                      |
  | You should have received a copy of the GNU General General Public    |
  | License in the file LICENSE along with this library;                 |
  | if not, write to the                                                 |
  |                                                                      |
  |   Free Software Foundation, Inc.,                                    |
  |   51 Franklin Street, Fifth Floor,                                   |
  |   Boston, MA  02110-1301  USA                                        |
  +----------------------------------------------------------------------+
  | Authors: Hartmut Holzgraefe <hartmut@mariadb.com>                    |
  +----------------------------------------------------------------------+
*/


/* TODO
 * query counter?
 * track first, last queries?
 * retrieve connection annotations?
 */


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <limits.h>
#include <time.h>

#include <mysql_version.h>

#include <mysql/plugin_audit.h>

#include <mysql/services.h>

#define SESSION_VAR(THD, VAR) (((struct my_vars *)THDVAR(THD, session_data))->VAR)

/* log long running transactions that took more than this many seconds */
static MYSQL_THDVAR_ULONGLONG(limit,
                              PLUGIN_VAR_RQCMDARG,
                              "log transactions that took longer than this many seconds",
                              /* check */ NULL,
                              /* update */ NULL,
                              /* default*/ 10,
                              /* min */ 0,
                              /* max */ ULLONG_MAX,
                              /* blocksize */ 1
                              );

/* this is an internal per-connection structure only,
 so NOCMDOPT and NOSYSVAR hide it in --help and 
 SHOW VARIABLES output */

static MYSQL_THDVAR_ULONGLONG(session_data,
                              PLUGIN_VAR_READONLY | PLUGIN_VAR_NOCMDOPT | PLUGIN_VAR_NOSYSVAR,
                              "private structure pointer",
                              /* check */ NULL,
                              /* update */ NULL,
                              /* default*/ 0,
                              /* min */ 0,
                              /* max */ ULLONG_MAX,
                              /* blocksize */ 1
                              );

/* register session variables declared above */
static struct st_mysql_sys_var* audit_plugin_sysvars[] = {
  MYSQL_SYSVAR(limit),
  MYSQL_SYSVAR(session_data),
  NULL
};

/* private session data structure */
struct my_vars {
  MYSQL_XID xid;
  char *user;
  time_t start_time;
};

/* SHOW_FUNC callback function that returns time the transaction started */
static int show_transaction_start(MYSQL_THD thd,
                                      struct st_mysql_show_var* var,
                                      char *buff)
{
  unsigned long long *result = (unsigned long long *)buff;

  var->type  = SHOW_LONGLONG;
  var->value = buff;
  
  *result = (unsigned long long)SESSION_VAR(thd, start_time);

  return 0;
}

/* SHOW_FUNC callback function that returns time elapsed since session start */
static int show_transaction_time(MYSQL_THD thd,
                                      struct st_mysql_show_var* var,
                                      char *buff)
{
  unsigned long long *result = (unsigned long long *)buff;

  var->type  = SHOW_LONGLONG;
  var->value = buff;
  
  if (SESSION_VAR(thd, start_time)) {
    *result = (unsigned long long)(time(NULL) - SESSION_VAR(thd, start_time));
  } else {
    *result = 0;
  }

  return 0;
}

/* register SHOW STATUS variables and callbacks */
static struct st_mysql_show_var audit_plugin_statvars[]=
  {
    {"transaction_time_start", (char *)&show_transaction_start, SHOW_FUNC},
    {"transaction_time_span", (char *)&show_transaction_time, SHOW_FUNC},
    {NULL, NULL, SHOW_UNDEF}
  };

/* called whenever we detect the start of a new transaction 
 *   
 * note that transactions actually start on the first real statement,
 * not on BEGIN / START TRANSACTION
 */
static void begin_transaction(MYSQL_THD thd, const struct mysql_event_general *event)
{
  /* populate session vars, 
     right now only 'user' is stored, 
     xid is initialized to zero by calloc already ... */
  if (THDVAR(thd, session_data)) {
    SESSION_VAR(thd, user) = (char *)calloc(1, event->general_user_length + 1);
    memcpy(SESSION_VAR(thd, user), event->general_user, event->general_user_length);

    /* remember transaction start time */
    SESSION_VAR(thd, start_time) = event->general_time;
  }
}

/* called whenever we detect that a transaction ended */
static void end_transaction(MYSQL_THD thd) 
{
	/* simple logging */
  if (THDVAR(thd, session_data) && SESSION_VAR(thd, user)) {
		time_t t = time(NULL);
		struct tm lt = *localtime(&t);
    unsigned long time_taken = (unsigned long)(time(NULL) - SESSION_VAR(thd, start_time)); 

   if ((THDVAR(thd, limit)) && (time_taken >= THDVAR(thd, limit))) {
      fprintf(stderr, "%2.2d%2.2d%2.2d %2.2d:%2.2d:%2.2d [note] long transaction: user: '%s', time taken: %lu\n", 
							lt.tm_year % 100, lt.tm_mon, lt.tm_mday, lt.tm_hour, lt.tm_min, lt.tm_sec,
                SESSION_VAR(thd, user), time_taken);
    }
    SESSION_VAR(thd, start_time) = 0;
  }

}

/* the actual audit notify callback */
static void audit_notify(MYSQL_THD thd, unsigned int event_class, const void *event)
{   
  /* act depending on event class */
  switch (event_class) {

  /* client connect / disconnect */
  case MYSQL_AUDIT_CONNECTION_CLASS:
    {
      /* convert to correct event data pointer */
      const struct mysql_event_connection *event_connection=
        (const struct mysql_event_connection *) event;

      switch (event_connection->event_subclass) {
        
        // new connection / session: allocate and initialize session data
      case MYSQL_AUDIT_CONNECTION_CONNECT:  
        THDVAR(thd, session_data) = (unsigned long long)calloc(sizeof(struct my_vars), 1); 
        if (THDVAR(thd, session_data)) {
          SESSION_VAR(thd, xid).formatID = -1;
          SESSION_VAR(thd, start_time) = 0;
        }
        break;

        // connection ended: clean up
      case MYSQL_AUDIT_CONNECTION_DISCONNECT:
        end_transaction(thd);
        if (THDVAR(thd, session_data)) {
          if (SESSION_VAR(thd, user)) {
            free(SESSION_VAR(thd, user));
          }
          free((void *)THDVAR(thd, session_data));
        }
        break;
      }
    }
    break;

  /* handle events generated by SQL processing */
  case MYSQL_AUDIT_GENERAL_CLASS:
    {
      MYSQL_XID xid, *old_xid;

      /* convert to correct event data pointer */
      const struct mysql_event_general *event_general=
        (const struct mysql_event_general *) event;
      
      /* we don't distinguish between the various event
         subtypes here, we're just tracking XID value
         changes and act accordingly */
      if (THDVAR(thd, session_data)) {
        // get current and previous transaction IDs
        thd_get_xid(thd, &xid);
        old_xid = &SESSION_VAR(thd, xid);

        // a transaction is valid if its formatID is >= 0

        if (xid.formatID == -1) {
          // currently not in a valid transaction
          if (old_xid->formatID >= 0) {
            // transition from valid to invalid transaction ID -> transaction ended
            end_transaction(thd);
          }
        } else {
          // currently in a valid transaction
          if (old_xid->formatID == -1) {
            // previously not in a transaction -> transaction started
            begin_transaction(thd, event_general);
          } else if (memcmp(old_xid, &xid, sizeof(MYSQL_XID))) {
            // previously in a valid transaction, too, and transaction ID changed
            // old transaction ended, new one started
            end_transaction(thd);
            begin_transaction(thd, event_general);
          }
          // else: still in same transaction
        }

        // remember current transaction id
        memcpy(old_xid, &xid, sizeof(MYSQL_XID));
      } 
    }
    break;
  }

} 

/* audit specific plugin setup */
static struct st_mysql_audit audit_descriptor=
  {
    MYSQL_AUDIT_INTERFACE_VERSION,
    NULL,
    audit_notify,
    { MYSQL_AUDIT_GENERAL_CLASSMASK | MYSQL_AUDIT_CONNECTION_CLASSMASK }
  };

/* general plugin setup */
mysql_declare_plugin(audit)
{
  MYSQL_AUDIT_PLUGIN,
    &audit_descriptor, 
    "transaction_time",
    "Hartmut Holzgraefe <hartmut@mariadb.com>",
    "A simple audit plugin measuring transaction times",
    PLUGIN_LICENSE_GPL,
    NULL,
    NULL,
    0x0001,
    audit_plugin_statvars,
    audit_plugin_sysvars,
    NULL, /* placeholder for command line options, not available yet */
    }
mysql_declare_plugin_end;

#ifdef MARIA_PLUGIN_INTERFACE_VERSION

/* general plugin setup, MariaDB version */
maria_declare_plugin(audit)
{
  MYSQL_AUDIT_PLUGIN,
    &audit_descriptor, 
    "transaction_time",
    "Hartmut Holzgraefe <hartmut@mariadb.com>",
    "A simple audit plugin measuring transaction times",
    PLUGIN_LICENSE_GPL,
    NULL,
    NULL,
    0x0001,
    audit_plugin_statvars,
    audit_plugin_sysvars,
    "0.1", /* version string */
    MariaDB_PLUGIN_MATURITY_BETA,
    }
maria_declare_plugin_end;

#endif
/*
 * Local variables:
 * tab-width: 2
 * c-basic-offset: 2
 * End:
 * vim600: noet sw=2 ts=2 fdm=marker
 ** vim<600: noet sw=2 ts=2
 */
